src/main/java/org/zalando/nakadi/service/kpi/publisher/EventsProcessor.java

```diff
@@ -0,0 +1,119 @@
+                           @Value("${nakadi.kpi.events-queue-size:100}") final int eventsQueueSize) {
+        this.eventPublisher = eventPublisher;
+        this.uuidGenerator = uuidGenerator;
+        this.featureToggleService = featureToggleService;
+        this.batchCollectionTimeout = batchCollectionTimeout;
+        this.batchSize = batchSize;
+        this.pollTimeout = pollTimeout;
+        this.eventsQueueSize = eventsQueueSize;
+        this.eventTypeEvents = new ConcurrentHashMap<>();
+        this.executorService = Executors.newFixedThreadPool(workers);
+    }
+
+    private void sendEventBatch(final String etName) {
+        LOG.trace("Collecting batch for {} on {}", etName, Thread.currentThread().getName());
+        try {
+            final BlockingQueue<JSONObject> events = eventTypeEvents.get(etName);
+            if (events == null) {
+                return;
+            }
+
+            final long finishAt = System.currentTimeMillis() + batchCollectionTimeout;
+            int eventsCount = 0;
+            JSONArray jsonArray = null;
+            while (eventsCount != batchSize && System.currentTimeMillis() < finishAt) {
+                try {
+                    final JSONObject event = events.poll(pollTimeout, TimeUnit.MILLISECONDS);
+                    if (event != null) {
+                        if (jsonArray == null) {
+                            jsonArray = new JSONArray();
+                        }
+                        jsonArray.put(event);
+                        eventsCount++;
+                    }
+                } catch (final InterruptedException ie) {
+                    Thread.currentThread().interrupt();
+                }
+            }
+            try {
+                if (eventsCount == 0) {
+                    LOG.trace("No kpi events send to {}", etName);
+                    return;
+                }
+                eventPublisher.publish(jsonArray.toString(), etName, null);
+                LOG.trace("Published batch of {} to {}", eventsCount, etName);
+            } catch (final Exception e) {
+                LOG.error("Error occurred while publishing events to {}, {}", etName, e.getMessage(), e);
+            }
+        } finally {
+            if (featureToggleService.isFeatureEnabled(FeatureToggleService.Feature.KPI_COLLECTION)) {
+                executorService.submit(() -> sendEventBatch(etName));
+            }
+        }
+    }
+
+    final void enrichAndSubmit(final JSONObject event, final String etName) {
+        final JSONObject metadata = new JSONObject()
+                .put("occurred_at", Instant.now())
+                .put("eid", uuidGenerator.randomUUID());
+        event.put("metadata", metadata);
+
+        final BlockingQueue<JSONObject> events =
+                eventTypeEvents.computeIfAbsent(etName, etn ->
+                {
+                    LOG.trace("Schedule events collection");
+                    executorService.submit(() -> sendEventBatch(etName));
```

Comment:
I think there will be a problem if `sendEventBatch(etName)` will be called before the value is put in the map

Reply:
what kind of problem ?

Reply:
I think @adyach covered this case in line 62, or not? I mean, if `eventTypeEvents.get(etName)` returns `null` because the executor was faster than this code, than it returns and runs again on line 95 `executorService.submit(() -> sendEventBatch(etName))`.

But I would avoid writing more concurrent code if we already have a buffer and queue feature in KafkaProducer. I would try to leverage it the most because then we don't need to fine tune configuration for 2 buffers. All optimizations we do to `KafkaProducer` in term of buffer size and other params would also benefit this feature.

Reply:
ah ok, as @rcillo mentioned - there is a check for null. So everything will work fine.